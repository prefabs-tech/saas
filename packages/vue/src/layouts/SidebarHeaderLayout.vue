<template>
  <Layout
    :menu="menu"
    :no-locale-switcher="noLocaleSwitcher"
    :user-menu-location="userMenuLocation"
  >
    <template #header>
      <slot name="header"></slot>
    </template>

    <template #locales>
      <slot name="locales"></slot>
    </template>

    <template #userMenu>
      <UserMenu v-if="showUserMenu" />
    </template>

    <template #afterNavLinks>
      <slot name="afterNavLinks"></slot>
    </template>

    <template #afterSidebarMenu>
      <slot name="afterSidebarMenu"></slot>
    </template>

    <template #footer>
      <slot name="footer"></slot>
    </template>

    <slot></slot>
  </Layout>
</template>

<script lang="ts">
export default {
  name: "SidebarHeaderLayout",
};
</script>

<script setup lang="ts">
import { useConfig } from "@dzangolab/vue3-config";
import { SidebarHeaderLayout as Layout } from "@dzangolab/vue3-layout";
import { storeToRefs } from "pinia";
import { computed } from "vue";
import { useRouter } from "vue-router";

import UserMenu from "../components/UserMenu.vue";
import useUserStore from "../store";

import type { MenuItem, SidebarMenu } from "@dzangolab/vue3-layout";

defineProps({
  noLocaleSwitcher: Boolean,
  showUserMenu: {
    default: true,
    required: false,
    type: Boolean,
  },
  userMenuLocation: {
    default: "sidebar",
    type: String,
    validator: (value: string) => ["header", "sidebar"].includes(value),
  },
});

const userStore = useUserStore();

const { user } = storeToRefs(userStore);

const { layout: layoutConfig } = useConfig();

const router = useRouter();

const allRoutes = router.getRoutes();

const menu = computed(() => {
  let menuItems = layoutConfig?.mainMenu || [];

  if (!user.value) {
    menuItems = menuItems?.filter((item: MenuItem) => {
      const route = allRoutes.find((r) => {
        if (r.name === item.route && !r.meta?.authenticated) {
          return true;
        }

        if (item.children?.length) {
          return (item.children as MenuItem[]).some((child: MenuItem) => {
            return allRoutes.some(
              (cr) => cr.name === child.route && !cr.meta?.authenticated,
            );
          });
        }

        return false;
      });

      return !!route;
    }) as MenuItem[];
  }

  return menuItems.map((item: MenuItem) => {
    return {
      hide: item?.hide,
      icon: item?.icon,
      name: item.name,
      routeName: item.route,
      shortName: item?.shortName,
      children: item?.children?.map((childItem: MenuItem) => {
        return {
          hide: childItem?.hide,
          icon: childItem?.icon,
          name: childItem.name,
          routeName: childItem.route,
          shortName: childItem?.shortName,
        };
      }),
    };
  }) as SidebarMenu[];
});
</script>

<style lang="css">
.sidebar-header-layout > .sidebar .user-menu > ul {
  flex-direction: column;
}

.sidebar-header-layout > .sidebar .user-menu > ul > li,
.sidebar-header-layout > .sidebar .user-menu-dropdown {
  --_font-size: var(--font-size-min, 0.8rem);
  --_font-weight: var(--font-weight, 450);
  --dropdown-container-bg-color: #0870e5;
  --menu-highlight-color: #0870e5;

  font-size: var(--_font-size);
  font-weight: var(--_font-weight);
  width: 100%;
}

.sidebar-header-layout
  > .sidebar
  .user-menu
  > ul
  > li:has(.router-link-exact-active) {
  background-color: #0870e5;
}

.sidebar-header- > .sidebar .user-menu:not(.user-menu-dropdown) > ul > li > a,
.sidebar-header-layout > .sidebar .user-menu-dropdown,
.sidebar-header-layout > .sidebar .user-menu-dropdown > ul.dropdown > li {
  --_height: var(--nav-menu-height, 3rem);

  align-items: center;
  display: flex;
  height: var(--_height);
  width: 100%;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown:hover {
  background-color: #0870e5;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown.expanded > ul > li:hover {
  --dropdown-bg-color-hover: #0870e5;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown.expanded .toggle > svg {
  transform: rotate(-180deg);
}

.sidebar-header-layout > .sidebar .user-menu-dropdown .toggle > svg {
  transform: rotate(0);
}

.sidebar-header-layout > .sidebar .user-menu-dropdown > .dropdown {
  opacity: 0;
  transform: translate3d(0, 0, 0);
  transition:
    transform 0.3s ease,
    opacity 0.5s ease;
}

.sidebar-header-layout > .sidebar .user-menu-dropdown.expanded > .dropdown {
  --dropdown-bg-color: #007aff;
  --dropdown-border: 1px solid #007aff;

  box-shadow: 0 -2px 10px 2px #0870e5;
  opacity: 1;
  transform: translate3d(0, -152.5px, 0);
  width: 100%;
}
</style>
