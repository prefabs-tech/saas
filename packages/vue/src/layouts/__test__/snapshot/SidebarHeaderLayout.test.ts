import configPlugin from "@dzangolab/vue3-config";
import { mount } from "@vue/test-utils";
import { createPinia } from "pinia";
import { describe, expect, it } from "vitest";
import { createRouter, createWebHistory } from "vue-router";

import SidebarHeaderLayout from "../../SidebarHeaderLayout.vue";
import appConfig from "../config";

import type { VueWrapper } from "@vue/test-utils";

describe("SidebarHeaderLayout", () => {
  it("matches snapshot", () => {
    const pinia = createPinia();
    const router = createRouter({
      history: createWebHistory(),
      routes: [
        {
          path: "/",
          name: "home",
          component: SidebarHeaderLayout,
        },
      ],
    });

    const wrapper: VueWrapper = mount(SidebarHeaderLayout, {
      global: {
        plugins: [
          pinia,
          [
            configPlugin,
            {
              config: appConfig,
            },
          ],
          router,
        ],
        stubs: {
          UserMenu: true,
          LocaleSwitcher: true,
        },
      },
      slots: {
        afterNavLinks: "<div class='after-nav'></div>",
        afterSidebarMenu: "<div class='after-sidebar'></div>",
        default: "<div class='default'></div>",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
