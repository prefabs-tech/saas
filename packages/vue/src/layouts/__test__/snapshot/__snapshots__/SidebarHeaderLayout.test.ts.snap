// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarHeaderLayout > matches snapshot 1`] = `
<div
  class="layout sidebar-header-layout"
>
  
  <header>
    
    <!---->
    
    <nav
      data-expanded="false"
    >
      
      <!---->
      
      
      
      
      
      
      <locale-switcher-stub
        class="locales"
      />
      
    </nav>
    <div
      class="toggle"
    >
      
      <svg
        aria-hidden="true"
        height="1.5rem"
        role="img"
        viewBox="0 0 16 16"
        width="1.5rem"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      />
      
    </div>
  </header>
  
  <div
    class="no-sidebar layout-sidebar"
  >
    <img
      alt="show sidebar"
      src="data:image/svg+xml,%3csvg%20viewBox='0%200%2024%2024'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cg%20stroke-width='0'%3e%3c/g%3e%3cg%20stroke-linecap='round'%20stroke-linejoin='round'%3e%3c/g%3e%3cg%3e%3cpath%20fill-rule='evenodd'%20clip-rule='evenodd'%20d='M8.29289%204.29289C8.68342%203.90237%209.31658%203.90237%209.70711%204.29289L16.7071%2011.2929C17.0976%2011.6834%2017.0976%2012.3166%2016.7071%2012.7071L9.70711%2019.7071C9.31658%2020.0976%208.68342%2020.0976%208.29289%2019.7071C7.90237%2019.3166%207.90237%2018.6834%208.29289%2018.2929L14.5858%2012L8.29289%205.70711C7.90237%205.31658%207.90237%204.68342%208.29289%204.29289Z'%20fill='%23ffffff'%3e%3c/path%3e%3c/g%3e%3c/svg%3e"
    />
  </div>
  <main>
    
    
    <div
      class="default"
    />
    
    
  </main>
</div>
`;
